### PR TITLE
[MCC-201759] Redesigned loader APIs and added dynamic column feature

### PR DESCRIPTION
**Pull request by junshao**
- Added dynamic column feature.
- Uses sheet definition models instead of sheet definition interfaces.
- Defines sheet name and column names by `SheetNameAttribute` and `ColumnHeaderNameAttribute` on sheet definition models.
- Redesign loader APIs.
- Removed dependency of `ImpromptuInterface` (no `.ActLike()` anymore).
- Cleaned up dependencies and code.
- Uploaded packages of `Medidata.Rave.Tsdv.Loader 0.1.3` and `Medidata.Cloud.ExcelLoader 0.1.3`.
- Added `unity.config` in sample code project, that Rave codebase can reuse.
- See below sample code. Refer to "Medidata.Rave.Tsdv.Loader.Sample/Program.cs" for more details.
   ```cs
var loader = container.Resolve<IExcelLoader>();

// Case 1
// Define a sheet by model type, and add items
loader.Sheet<BlockPlan>();
loader.Sheet<BlockPlan>().Data.Add(
    new BlockPlan
    {
        BlockPlanName = "xxx",
        UsingMatrix = false,
        EstimatedDate = DateTime.Now,
        EstimatedCoverage = 0.85
    },
    new BlockPlan {BlockPlanName = "yyy", EstimatedCoverage = 0.65},
    new BlockPlan {BlockPlanName = "zzz"});

// Case 2
// Automatically define sheet when initially calling SheetData with new type
loader.Sheet<BlockPlanSetting>().Data.Add(
    new BlockPlanSetting
    {
        BlockPlanName = "fakeNameByAnonymousClass",
        Repeated = false,
        BlockSubjectCount = 99
    },
    new BlockPlanSetting {BlockPlanName = "111", Repeated = true, BlockSubjectCount = 100},
    new BlockPlanSetting {BlockPlanName = "ccc", Blocks = "fasdf"});

// Case 3
// Add dynamic columns and add extra properties to model object.
loader.Sheet<TierFolder>().Definition
        .AddColumn(new ColumnDefinition
                    {
                        PropertyName = "Visit1",
                        PropertyType = typeof(bool)
                    })
        .AddColumn(new ColumnDefinition
                    {
                        PropertyName = "Visit2",
                        PropertyType = typeof(int)
                    })
        .AddColumn(new ColumnDefinition
                    {
                        PropertyName = "Unscheduled",
                        PropertyType = typeof(string)
                    });

loader.Sheet<TierFolder>().Data.Add(
    new TierFolder {TierName = "T1", FolderOid = "FOLDETR"}.AddProperty("Visit1", true),
    new TierFolder {TierName = "T1", FolderOid = "FOLDETR"}.AddProperty("Visit2", 1),
    new TierFolder {TierName = "T1", FolderOid = "FOLDETR"}.AddProperty("Unscheduled", "x"));

var filePathX = @"C:\Github\test.xlsx";
File.Delete(filePathX);
using (var fs = new FileStream(filePathX, FileMode.Create))
{
    loader.Save(fs);
}

// Use parser to load a .xlxs file
using (var fs = new FileStream(filePathX, FileMode.Open))
{
    loader.Load(fs);
}

Console.WriteLine(loader.Sheet<BlockPlan>().Data.First().BlockPlanName);
Console.WriteLine(loader.Sheet<BlockPlanSetting>().Data.Count);
// Load extra properties from extra columns.
Console.WriteLine(loader.Sheet<TierFolder>().Data.First().GetExtraProperties()["Visit1"]);
Console.WriteLine(loader.Sheet<Rule>().Data.Count);
```
@art2003 @echen-mdsol @chenghuang-mdsol Please review and merge if OK. Thanks.

### CODE REVIEW

File: Medidata.Cloud.ExcelLoader/ExcelBuilder.cs [hunk 1]
```diff
@@ -1,41 +1,26 @@
...
-        private readonly IList<ISheetBuilder> _sheetBuilders = new List<ISheetBuilder>();
+        private readonly IDictionary<string, SheetModels> _modelDic = new Dictionary<string, SheetModels>();
 
-        public ExcelBuilder(ISheetBuilderFactory sheetBuilderFactory)
+        public void AddSheet(ISheetDefinition sheetDefinition, IEnumerable<SheetModel> models,
+                             ISheetBuilder sheetBuilder)
         {
-            if (sheetBuilderFactory == null) throw new ArgumentNullException("sheetBuilderFactory");
-            _sheetBuilderFactory = sheetBuilderFactory;
+            var sheetName = sheetDefinition.Name;
```
Comment (art2003) [inline Medidata.Cloud.ExcelLoader/ExcelBuilder.cs]: Public method, should we check for parameters existence

Comment (junshao) [inline Medidata.Cloud.ExcelLoader/ExcelBuilder.cs]: Null check added. Thanks.

File: Medidata.Cloud.ExcelLoader/ExcelBuilder.cs [hunk 2]
```diff
@@ -50,39 +35,25 @@ public void Save(Stream outStream)
                     workbookPart = doc.WorkbookPart;
                 }
 
-                foreach (var sheet in _sheetBuilders)
+                foreach (var key in _modelDic.Keys)
                 {
-                    sheet.AttachTo(doc);
+                    var info = _modelDic[key];
+                    info.SheetBuilder.BuildSheet(info, info.SheetDefinition, doc);
```
Comment (art2003) [inline Medidata.Cloud.ExcelLoader/ExcelBuilder.cs]: If we don't use key for dictionary, can we just loop over and get value?

 foreach (var entry in _modelDic)
            {
                var info = entry.Value;
            }

File: Medidata.Cloud.ExcelLoader/ExcelLoader.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        public virtual void Load(Stream source)
+        {
+            _parser.Load(source);
+
+            foreach (var type in _sheetInfoDic.Keys)
+            {
+                var info = _sheetInfoDic[type];
+                var sheetDef = info.SheetDefinition;
+                info.LoadedData = _parser.GetObjects(sheetDef, _sheetParser).ToList();
+            }
```
Comment (art2003) [inline Medidata.Cloud.ExcelLoader/ExcelLoader.cs]: similar as above

**Comment by junshao** [comment]
@art2003 My VM expired and cannot reflect your comments soon. Maybe we can merge this as it is, and I will open other PR to improve. Thanks.

**Comment by art2003** [comment]
ok, agree

**Comment by art2003** [comment]
merging

**Pull request merged**